### PR TITLE
Introduce theming policy API that makes theme switching possible

### PR DESCRIPTION
**Pull request by gyst**
If you google for "plone theme switching" you get ~ 41.000 results. It was a core feature in Plone in the 2.x series and well supported until Plone4 with collective.themeswitcher. However, changes introduced in the run-up to Plone5 have made theme switching impossible. I've found multiple partially overlapping theme caching implementations in p.a.theming that all hardcode the assumption that there is only one theme. Additionally, Products.CMFPlone messes with one of those caches as well (yuk), for which I will issue a separate pull request if this p.a.theming pull is accepted. All of this is hardcoded in such a way that it cannot be reached by overrides or even monkey patches.

The full scope of this pull request is as follows:
- This PR consolidates all theme choosing decisions into a policy facade.
- All existing API entry points relating to this are mapped to the new policy facade in a backward-compatible way.
- The policy framework here does *not* introduce theme switching. That's out of scope. It merely provides a backward- and forward-compatible foundation for doing theme switching outside of core.
- The policy framework introduced works as a request adapter that can be easily swapped by integrators using normal configure.zcml registrations bound to a browser layer, not even requiring an overrides.zcml.
- The policy maintains a single cache manager that supports multiple theme caches per site. Instead of using a _v_volatile it uses a class static to store the cache manager. This is needed because a _v_volatile introduces CSRF write-on-read errors if you actually start using theme switching outside of core. This has a slight performance impact in that the theme needs to be compiled once per Plone instance start, instead of just once per site.

Outside of this PR you can see how it makes theme switching possible in the Plone Intranet themeswitcher implementation.
- documentation: http://docs.ploneintranet.org/development/components/themeswitcher.html
- code: https://github.com/ploneintranet/ploneintranet/tree/master/src/ploneintranet/themeswitcher

Note that ploneintranet.themeswitcher is a separate package. It has it's own configure.zcml, it's own generic setup and it's own test suite. It can be easily factored out of ploneintranet and I'm happy to support that if there's an interest. For now it just implements the minimal set we need for Plone Intranet - I'm not sure whether it makes sense to support CMF skin switching etc in a Plone5 world and there would need to be some extra discussion I guess to see whether/how it makes sense to support other collective.themeswitcher features.

For theme switching to actually work, CMFPlone needs to stop messing around with theme caching internals. That will become a separate pull request.
https://github.com/ploneintranet/Products.CMFPlone/commit/815044838d7a8732ff69d740067d3ab2e8f39ae0
Since that change depends on the new policy adapter being available it only makes sense to issue a CMFPlone pull request once this p.a.theming pull has been merged.

### CODE REVIEW

File: src/plone/app/theming/policy.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        key = "%s::%s" % (getSite().absolute_url(), theme)
+        return key
+
+    def getCacheStorage(self):
+        return self.__class__.CACHE_STORAGE
+
+    def invalidateCache(self):
+        """When our settings are changed, invalidate the cache on all zeo clients
+        """
+        self.__class__.CACHE_STORAGE = {}
```
Comment (vangheem) [inline src/plone/app/theming/policy.py]: How do threads deal with class attributes? Should we be using a cache stored in a thread local?

File: src/plone/app/theming/policy.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+    def get_theme(self):
+        """Managing the theme cache is a plone.app.theming policy
+        decision. Moved out out Products.CMFPlone."""
+        cache = self.getCache()
+        themeObj = cache.themeObj
+        if not themeObj:
+            theme = self.getCurrentTheme()
+            themeObj = utils.getTheme(theme)
+            self.set_theme(theme, themeObj)
```
Comment (vangheem) [inline src/plone/app/theming/policy.py]: good idea to use set in a get?

**Comment by vangheem** [comment]
I made a couple comments on the code. Thanks for looking into this.

Does it make it any easier if you go back to using _v_ volatile attributes? plone.protect allows you to use _v_safe_write to specify writing an object is safe.

**Comment by vangheem** [comment]
I'm surprised _v_ would give you any trouble at all though really with write on read. _v_ specially means that it won't be stored in the database. _v_ attributes should only be use for instance caches so I'm not sure there will be much of any performance difference with your code. I think the registry._p_changed = True in the old code was unnecessary perhaps?

Thinking about this... How do theme changes affect caches on other instances(or even other threads)? It seems to me that this approach would easily produce inconsistencies across multiple clients.

Ideally, we have a theme modification date we pay attention to when pulling from cache. Cache would store when it was last generated, compare against theme modification date, etc. This of course only matters for TTW themes.

File: src/plone/app/theming/policy.py (hunk 2 above)
Comment (gyst) [inline src/plone/app/theming/policy.py]: I don't see an easy way around that. My use case is, that the fallback theme (barceloneta) is generated on the first hit on cms.localhost:8080. This is what causes the write-on-read errors and why I moved away from v_volatile.

**Comment by gyst** [comment]
Ah, I discarded the _v_safe_write scenario as an option when the cache was still on the registry, for security reasons. However there's no reason to not declare my own cache _v_safe_write. In combination with your suggestion to avoid _p_changed that might work. Will try that change.

**Comment by vangheem** [comment]
Your approach is still probably better. Use "_v_" on the registry was a bit wonky. I still don't understand how invalidation works that way with multiple threads or clients. Perhaps I'm not understanding some piece of it.

**Comment by gyst** [comment]
Yes I just reconsidered and really don't want to mark the registry as _v_safe_write and I have no other persistent object that I would consider compromising that way.

File: src/plone/app/theming/policy.py (hunk 1 above)
Comment (gyst) [inline src/plone/app/theming/policy.py]: I have no idea :-(. Any suggestions on how to bring that under test?

Comment (vangheem) [inline src/plone/app/theming/policy.py]: I don't. :( Using a thread local might be a good way to go to be safe anyways. Whenever I do this sort of caching, I always use thread locals.

Comment (gyst) [inline src/plone/app/theming/policy.py]: Do you have any good thread-local examples to get me started? Google is not very helpful.

Comment (vangheem) [inline src/plone/app/theming/policy.py]: It's basically like this:

```
import threading
_local_cache = threading.local()

...
def getCacheStorage(self):
    if not hasattr(_local_cache, 'data'):
        _local_cache.data = {}
    return _local_cache.data
```

Comment (gyst) [inline src/plone/app/theming/policy.py]: Thanks, will work on that.

**Comment by datakurre** [comment]
Just FYI that there was an alternative implementation, which allowed custom themes for subsites by layering / proxying registry:

* https://pypi.python.org/pypi/collective.behavior.localdiazo
* https://pypi.python.org/pypi/collective.behavior.localregistry

Yet, probably this is more explicit and therefore better (proxying is always a bit magic).

**Comment by datakurre** [comment]
About the old code. AFAIK ``registry._p_changed = True`` adds registry into transaction, which makes zeo to send invalidation message about cached registry objects to all zeo connections. And because volatile ``_v_`` attributes are saved with the cached object, they are thrown away and so theme cached gets purged on all zeo connections. 

http://zodborg.readthedocs.org/en/latest/documentation/articles/ZODB2.html#not-all-objects-are-persistent

**Comment by vangheem** [comment]
Ah hah, that makes sense. That would explain how it works with multiple
threads/clients. In that case, I think we need to stick with volatile
attributes.

**Comment by gyst** [comment]
Oh @datakurre @vangheem I see the above comments only after I implemented this :-).

So yes the cache is now stored in a thread local, while signalling to other threads and connections is done via a persistent marker. Because the lazy cache writes in my switched theme are in the thread local rather than in a CSRF protected _v_ attributes this avoids write-on-read errors, while the ZODB is still used for inter-process signalling of cache invalidation.

I've tested this manually in a ZEO setup.

**Comment by gyst** [comment]
@tisto do I need to do something special to get this pull request jenkins tested?

**Comment by datakurre** [comment]
Not sure how the permissions are managed, but after login (uses github login), a new buld can be triggered by pasting pull request url into http://jenkins.plone.org/job/pull-request-5.0/build

http://jenkins.plone.org/job/pull-request-5.0/39/

**Comment by gyst** [comment]
This was tested OK by Jenkins in http://jenkins.plone.org/job/pull-request-5.0/55/.
Don't understand why Github cannot automatically merge - I merged master into this branch and it's fully up to date.

I hope this can now be merged. Caching is thread local, but cache invalidation works across threads and processes, so that should be cool.

**Comment by vangheem** [comment]
Looks fine. I'll see about merging it.

**Comment by vangheem** [comment]
Maybe because you created a fork to another repository? You need to rebase from this original repository instead of your fork.

I have manually merged: https://github.com/plone/plone.app.theming/commit/3756dbd45db394f112fae9ff25a532934013e1b5

Can you verify it looks fine? You also forgot a changelog.

**Comment by gyst** [comment]
Thanks for the merge. I pushed and updated changelog directly to master.
Have now opened follow-up PR for CMFPlone:
https://github.com/plone/Products.CMFPlone/pull/667

**Pull request closed without merging**